src/QirRuntime/test/QIR-static/qir-test-math.qs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Testing.QIR.Math
+{
+    open Microsoft.Quantum.Intrinsic;
+    open Microsoft.Quantum.Math;        // E()
+
+    function SqrtTest() : Int {
+        if( 2.0 != Sqrt(  4.0) )    { return 1; }     
+        if( 3.0 != Sqrt(  9.0) )    { return 2; }     
+        if(10.0 != Sqrt(100.0) )    { return 3; }     
+
+        if( not IsNan(Sqrt(-5.0)))        { return 4; }      
+        if( not IsNan(Sqrt(NAN())))       { return 5; }      
+        if( not IsInf(Sqrt(INFINITY())))  { return 6; }      
+
+        return 0;
+    }
+
+    function LogTest() : Int {
+        if( 1.0 != Log(E()) )                       { return 1; }       // ln(e)   -> 1
+        if( 2.0 != Log(E() * E()) )                 { return 2; }       // ln(e^2) -> 2
+
+        if( not IsNegativeInfinity(Log(0.0)))       { return 3; }        // ln(0)           -> (-nfinity)
+        if( not IsNan(Log(-5.0)))                   { return 4; }        // ln(<negaive>)   -> NaN
+        if( not IsNan(Log(NAN())))                  { return 5; }        // ln(NaN)         -> NaN
+        if( not IsInf(Log(INFINITY())))             { return 6; }        // ln(+infinity)   -> +infinity
+
+        return 0;
+    }
+
+    function ArcTan2Test() : Int {
+
+        // function ArcTan2(y : Double, x : Double) : Double 
+
+        if(  0.0            != ArcTan2( 0.0,  1.0 ) )          { return 1; }
+        if(  PI()           != ArcTan2( 0.0, -1.0 ) )          { return 2; }
+        if(  PI()/2.0       != ArcTan2( 1.0,  0.0 ) )          { return 3; }
+        if( -PI()/2.0       != ArcTan2(-1.0,  0.0 ) )          { return 4; }
+
+        if(  PI()/4.0       != ArcTan2( 1.0,  1.0 ) )          { return 5; }
+        if(  PI()*3.0/4.0   != ArcTan2( 1.0, -1.0 ) )          { return 6; }
+        if( -PI()*3.0/4.0   != ArcTan2(-1.0, -1.0 ) )          { return 7; }
+        if( -PI()/4.0       != ArcTan2(-1.0,  1.0 ) )          { return 8; }
+
+        if(  0.0            != ArcTan2( 0.0,  0.0 ) )          { return 9; }    
+
+        // Fails because of lack of precision:
+        // if(  PI()/6.0  != ArcTan2( 1.0, Sqrt(3.0) ) )     { return 10; }      // tg(Pi/6) = sin(Pi/6) / cos(Pi/6) = (1/2) / (Sqrt(3)/2) = 1/Sqrt(3) = y/x.  ArcTan2(1.0, Sqrt(3)) -> Pi/6
+
+        if( not IsNan(ArcTan2(             NAN(),                0.0) )) { return 11; }
+        if( not IsNan(ArcTan2(               0.0,              NAN()) )) { return 12; }
+        if( not IsNan(ArcTan2(             NAN(),              NAN()) )) { return 13; }
+
+        // The infinity cases show discrepancy between  
+        // https://docs.microsoft.com/en-us/dotnet/api/system.math.atan2?view=net-5.0
```

Comment:
Looks like the [C#'s (.NET's) semantics](https://docs.microsoft.com/en-us/dotnet/api/system.math.atan2?view=net-5.0) somewhat contradicts to itself:
On the one hand it says: 
> if x and y are either PositiveInfinity or NegativeInfinity, the method returns NaN

On the other hand it remarks:
> This method calls into the underlying C runtime, and the exact result or valid input range may differ between different operating systems or architectures.

The [C](https://en.cppreference.com/w/c/numeric/math/atan2) (and [C++](https://en.cppreference.com/w/cpp/numeric/math/atan2)) runtime returns NaN in one case only - "If either x is NaN or y is NaN".

For now I suspect that the phrase 
> if x and y are either PositiveInfinity or NegativeInfinity, the method returns NaN

is out-of-date even for C# (and for the whole .NET), and even in C# the behavior corresponds to the C (and C++) runtime.
So for now I would leave the implementation and documentation for Q# as is (since Guen is blocked by this missing functionality), and get back to the discussion later if needed.

I filed a documentation issue [77](https://github.com/microsoft/qsharp-language/issues/77) in qsharp-language.

Reply:
Confirming, it looks like the current behavior of Q#'s `ArcTan2` follows what I would expect from the cppreference docs that you shared:

![image](https://user-images.githubusercontent.com/31516/107298952-40e7c080-6a2b-11eb-840d-6544ebf56444.png)

This would be good to test for explicitly in the C#-based runtime, but given that the two implementations seem to agree, I agree that it makes sense to defer adding those tests until the next PR.

Reply:
Thanks for confirming! Resolving this conversation for now (in this PR).